app/agent/runtime/qemu/qemu.service.ts

```diff
@@ -0,0 +1,542 @@
+import { spawn } from "child_process";
+import fs from "fs/promises";
+import { Socket } from "net";
+import path from "path";
+
+import type { DefaultLogger } from "@temporalio/worker";
+import { Netmask } from "netmask";
+import type { Config as SSHConfig, NodeSSH } from "node-ssh";
+import { match } from "ts-pattern";
+
+import type { AgentUtilService } from "../../agent-util.service";
+import { MAX_UNIX_SOCKET_PATH_LENGTH } from "../../constants";
+import { FifoFlags } from "../../fifo-flags";
+import { doesFileExist, execCmd, withSsh } from "../../utils";
+import type { HocusRuntime } from "../hocus-runtime";
+import type { VmInfo } from "../vm-info.validator";
+import { VmInfoValidator } from "../vm-info.validator";
+
+import type { IpBlockId, WorkspaceNetworkService } from "~/agent/network/workspace-network.service";
+import { NS_PREFIX } from "~/agent/network/workspace-network.service";
+import type { Config } from "~/config";
+import type { PerfService } from "~/perf.service.server";
+import { Token } from "~/token";
+import { displayError, sleep, unwrap, waitForPromises } from "~/utils.shared";
+
+export class QmpConnectionFailedError extends Error {}
+
+const CHROOT_PATH_TO_SOCK = "/run/sock";
+// https://github.com/torvalds/linux/blob/8b817fded42d8fe3a0eb47b1149d907851a3c942/include/uapi/linux/virtio_blk.h#L58
+const VIRTIO_BLK_ID_BYTES = 20;
+
+export function factoryQemuService(
+  agentUtilService: AgentUtilService,
+  logger: DefaultLogger,
+  config: Config,
+  perfService: PerfService,
+  networkService: WorkspaceNetworkService,
+): (instanceId: string) => HocusRuntime {
+  return (instanceId: string) =>
+    new QemuService(agentUtilService, logger, config, perfService, networkService, instanceId);
+}
+factoryQemuService.inject = [
+  Token.AgentUtilService,
+  Token.Logger,
+  Token.Config,
+  Token.PerfService,
+  Token.WorkspaceNetworkService,
+] as const;
+
+export class QemuService implements HocusRuntime {
+  private readonly qmpSocketPath: string;
+  private readonly instanceDir: string;
+  private readonly instanceDirRoot: string;
+  private readonly vmInfoFilePath: string;
+  private readonly agentConfig: ReturnType<Config["agent"]>;
+
+  constructor(
+    private readonly agentUtilService: AgentUtilService,
```

Comment:
this service isn't using agentUtilService, config, and perfService, so they can be removed